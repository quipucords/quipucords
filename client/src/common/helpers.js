export const bindMethods = (context, methods) => {
  methods.forEach(method => {
    context[method] = context[method].bind(context);
  });
};

export const noop = Function.prototype;
