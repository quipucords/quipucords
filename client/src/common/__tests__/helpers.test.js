import { helpers } from '../helpers';

describe('Helpers', () => {
  it('should have specific functions', () => {
    expect(helpers).toMatchSnapshot('helpers');
  });

  it('should support generated strings and flags', () => {
    expect(helpers.generateId()).toBe('generatedid-');
    expect(helpers.generateId('lorem')).toBe('lorem-');

    expect(helpers.FULFILLED_ACTION('lorem')).toBe(`lorem_FULFILLED`);
    expect(helpers.PENDING_ACTION('lorem')).toBe(`lorem_PENDING`);
    expect(helpers.REJECTED_ACTION('lorem')).toBe(`lorem_REJECTED`);
  });

  it('should support icon references', () => {
    expect(helpers.sourceTypeIcon('vcenter')).toMatchSnapshot('sourceTypeIcon');
    expect(helpers.scanTypeIcon('connect')).toMatchSnapshot('scanTypeIcon');
    expect(helpers.scanStatusIcon('success')).toMatchSnapshot('scanStatusIcon');
  });

  it('should handle basic object updates', () => {
    const definedObj = {};
    expect(helpers.setPropIfDefined(definedObj, ['lorem'], undefined)).toMatchSnapshot('setPropIfDefined undefined');
    expect(helpers.setPropIfDefined(definedObj, ['lorem'], 'ipsum')).toMatchSnapshot('setPropIfDefined');

    const truthyObj = {};
    expect(helpers.setPropIfTruthy(truthyObj, ['lorem'], 0)).toMatchSnapshot('setPropIfTruthy number');
    expect(helpers.setPropIfTruthy(truthyObj, ['lorem'], true)).toMatchSnapshot('setPropIfTruthy bool');
    expect(helpers.setPropIfTruthy(truthyObj, ['lorem'], '')).toMatchSnapshot('setPropIfTruthy string');
  });

  it('should update a state object', () => {
    const initialState = {
      lorem: false,
      ipsum: true
    };

    const state = {};
    state.ipsum = false;

    expect(
      helpers.setStateProp(
        null,
        {
          lorem: true
        },
        {
          state,
          initialState
        }
      )
    ).toMatchSnapshot('reset state object');

    state.ipsum = false;

    expect(
      helpers.setStateProp(
        null,
        {
          lorem: true
        },
        {
          state,
          reset: false
        }
      )
    ).toMatchSnapshot('dont reset state object');
  });

  it('should handle view related selectors and props updates', () => {
    expect(helpers.viewPropsChanged({ activeFilters: true }, { activeFilters: false })).toBe(true);

    const viewOptions = {
      currentPage: 1,
      pageSize: 10
    };

    const queryObj = {
      page: 2
    };

    expect(helpers.createViewQueryObject(viewOptions, queryObj)).toMatchSnapshot('createViewQueryObject');
  });

  it('should handle http status less than 400 message from response', () => {
    const payload = {
      data: {},
      status: 201,
      statusText: 'Created',
      headers: {},
      config: {},
      request: {},
      message: 'Request success XXX',
      detail: 'XXX Request success'
    };

    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('200 message');

    payload.status = 399;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('399 message');

    payload.status = 400;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('400 message');
  });

  it('should handle an undefined http status', () => {
    const payload = {
      message: 'MESSAGE',
      response: {
        status: 0,
        statusText: 'ERROR TEST',
        data: {
          detail: 'ERROR'
        }
      }
    };

    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('0 message');

    delete payload.response.status;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('undefined message');
  });

  it('should handle http status 400 level error messages from response', () => {
    const payload = {
      response: {
        data: {
          lorem: ['Lorem ipsum dolor sit']
        },
        status: 400,
        statusText: 'Bad Request',
        headers: {},
        config: {},
        request: {}
      },
      message: 'Request failed with status code XXX',
      detail: 'XXX Request failed'
    };

    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('400 error message');

    expect(helpers.getMessageFromResults(payload, 'lorem').message).toMatchSnapshot('400 filtered message');

    expect(helpers.getMessageFromResults(payload, 'dolor').message).toMatchSnapshot('400 filtered blank message');

    delete payload.response.data.lorem;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('400 fallback error message');

    delete payload.response.data;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('400 fallback again error message');

    delete payload.message;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('400 fallback yet again error message');
  });

  it('should handle http status 500 level error messages from response', () => {
    const payload = {
      response: {
        data: {
          lorem: ['Lorem ipsum dolor sit']
        },
        status: 500,
        statusText: 'Bad Request',
        headers: {},
        config: {},
        request: {}
      },
      message: 'Request failed with status code XXX',
      detail: 'XXX Request failed'
    };

    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('500 level error message');

    delete payload.response.data.lorem;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('500 level fallback error message');

    delete payload.response.data;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot('500 level fallback again error message');

    delete payload.message;
    expect(helpers.getMessageFromResults(payload).message).toMatchSnapshot(
      '500 level fallback yet again error message'
    );
  });

  it('should return http status from a response', () => {
    const payload = {
      response: {
        status: 400
      }
    };

    expect(helpers.getStatusFromResults({})).toBe(0);

    expect(helpers.getStatusFromResults(payload)).toBe(400);
  });

  it('should return timestamp from a response', () => {
    const payload = {
      headers: {
        date: 'Mon, 25 Feb 2019 21:46:40 GMT'
      }
    };

    expect(helpers.getTimeStampFromResults(payload)).toBe('20190225_164640');
  });

  it('should determine an ip address and value', () => {
    expect(helpers.isIpAddress('0.0.0.1')).toBe(true);
    expect(helpers.isIpAddress('0.0.0.1.5')).toBe(false);
    expect(helpers.isIpAddress('lorem')).toBe(false);

    expect(helpers.ipAddressValue('0.0.0.1')).toBe(1);
    expect(helpers.ipAddressValue('0.0.0.1.5')).toBe(1);
    expect(Number.isNaN(helpers.ipAddressValue('lorem'))).toBe(true);
  });
});
