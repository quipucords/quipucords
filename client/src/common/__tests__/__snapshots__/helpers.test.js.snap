// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0 message 1`] = `"ERROR"`;

exports[`200 message 1`] = `"Request success XXX"`;

exports[`399 message 1`] = `"Request success XXX"`;

exports[`400 error message 1`] = `"Lorem ipsum dolor sit"`;

exports[`400 fallback again error message 1`] = `"Request failed with status code XXX"`;

exports[`400 fallback error message 1`] = `""`;

exports[`400 fallback yet again error message 1`] = `"XXX Request failed"`;

exports[`400 filtered blank message 1`] = `""`;

exports[`400 filtered message 1`] = `"Lorem ipsum dolor sit"`;

exports[`400 message 1`] = `"Request success XXX"`;

exports[`500 level error message 1`] = `"500 Lorem ipsum dolor sit"`;

exports[`500 level fallback again error message 1`] = `"500 Request failed with status code XXX"`;

exports[`500 level fallback error message 1`] = `"500 "`;

exports[`500 level fallback yet again error message 1`] = `"500 XXX Request failed"`;

exports[`createViewQueryObject 1`] = `
Object {
  "page": 1,
  "page_size": 10,
}
`;

exports[`dont reset state object 1`] = `
Object {
  "ipsum": false,
  "lorem": true,
}
`;

exports[`helpers 1`] = `
Object {
  "DEV_MODE": false,
  "FULFILLED_ACTION": [Function],
  "PENDING_ACTION": [Function],
  "REJECTED_ACTION": [Function],
  "RH_BRAND": false,
  "TEST_MODE": true,
  "createViewQueryObject": [Function],
  "devModeNormalizeCount": [Function],
  "downloadData": [Function],
  "generateId": [Function],
  "getMessageFromResults": [Function],
  "getStatusFromResults": [Function],
  "getTimeStampFromResults": [Function],
  "ipAddressValue": [Function],
  "isIpAddress": [Function],
  "noop": [Function],
  "scanStatusIcon": [Function],
  "scanTypeIcon": [Function],
  "setPropIfDefined": [Function],
  "setPropIfTruthy": [Function],
  "setStateProp": [Function],
  "sourceTypeIcon": [Function],
  "viewPropsChanged": [Function],
}
`;

exports[`reset state object 1`] = `
Object {
  "ipsum": true,
  "lorem": true,
}
`;

exports[`scanStatusIcon 1`] = `
Object {
  "classNames": Array [],
  "name": "ok",
  "type": "pf",
}
`;

exports[`scanTypeIcon 1`] = `
Object {
  "name": "connected",
  "type": "pf",
}
`;

exports[`setPropIfDefined 1`] = `
Object {
  "lorem": "ipsum",
}
`;

exports[`setPropIfDefined undefined 1`] = `Object {}`;

exports[`setPropIfTruthy bool 1`] = `
Object {
  "lorem": true,
}
`;

exports[`setPropIfTruthy number 1`] = `Object {}`;

exports[`setPropIfTruthy string 1`] = `
Object {
  "lorem": true,
}
`;

exports[`sourceTypeIcon 1`] = `
Object {
  "name": "vcenter",
  "type": "pf",
}
`;

exports[`undefined message 1`] = `"ERROR"`;
